### PR TITLE
raft: hide Campaign rules on applying all entries

### PR DESCRIPTION
**Pull request by mrmiywj**
solve #873

### CODE REVIEW

**Comment by siddontang** [comment]
Thanks @mrmiywj 
Please sign the CLA first.

**Comment by siddontang** [comment]
Hi @mrmiywj The CI can't compile the code.

File: Cargo.lock [hunk 1]
```diff
@@ -580,3 +580,74 @@ dependencies = [
  "winapi-build 0.1.1 (registry+https://github.com/rust-lang/crates.io-index)",
 ]
 
+[metadata]
```
Comment (siddontang) [inline Cargo.lock]: How to generate it?
Seem we don't use this before.

**Comment by siddontang** [comment]
CI still can't compile. 😭

**Comment by mrmiywj** [comment]
@siddontang I'm not sure why it passed tests on linux but failed on OS X..

**Comment by siddontang** [comment]
seems that we should merge https://github.com/pingcap/tikv/pull/917 first.

File: Cargo.lock (hunk 1 above)
Comment (siddontang) [inline Cargo.lock]: @mrmiywj do we need to remove this?

File: src/raft/raft.rs [hunk 2]
```diff
@@ -602,18 +602,22 @@ impl<T: Storage> Raft<T> {
...
+
+        if nconf == 1 {
             self.pending_conf = true;
         }
         self.append_entry(&mut [Entry::new()]);
         info!("{} became leader at term {}", self.tag, self.term);
     }
 
+    fn num_pending_conf(&self, ents: &Vec<Entry>) -> usize {
+        ents.into_iter().filter(|e| e.get_entry_type() == EntryType::EntryConfChange).count()
```
Comment (siddontang) [inline src/raft/raft.rs]: should use iter instead? @BusyJay

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: src/raft/raft.rs [hunk 3]
```diff
@@ -664,6 +668,12 @@ impl<T: Storage> Raft<T> {
             if self.state == StateRole::Leader {
                 debug!("{} ignoring MsgHup because already leader", self.tag);
             } else {
+                let ents = self.raft_log.entries(self.raft_log.applied + 1, self.raft_log.committed - self.raft_log.applied).expect("unexpected error getting uncommitted entries");
```
Comment (BusyJay) [inline src/raft/raft.rs]: Why `self.raft_log.committed - self.raft_log.applied`?

File: src/raft/raft.rs [hunk 4]
```diff
@@ -664,6 +668,12 @@ impl<T: Storage> Raft<T> {
...
+                let n = self.num_pending_conf(&ents);
+                if n != 0 && self.raft_log.committed > self.raft_log.applied {
```
Comment (BusyJay) [inline src/raft/raft.rs]: if n is not 0, then committed must be larger than applied.

File: src/raft/raft.rs (hunk 2 above)
Comment (BusyJay) [inline src/raft/raft.rs]: `into_iter` is ok.

File: src/raft/raft.rs [hunk 5]
```diff
@@ -602,18 +602,22 @@ impl<T: Storage> Raft<T> {
...
+        }
+
+        if nconf == 1 {
             self.pending_conf = true;
         }
         self.append_entry(&mut [Entry::new()]);
         info!("{} became leader at term {}", self.tag, self.term);
     }
 
+    fn num_pending_conf(&self, ents: &Vec<Entry>) -> usize {
```
Comment (BusyJay) [inline src/raft/raft.rs]: Use `&[Entry]` instead.

File: src/raft/raft.rs (hunk 3 above)
Comment (mrmiywj) [inline src/raft/raft.rs]: This is added [here](https://github.com/coreos/etcd/pull/6039/files#diff-a8629a0eb6da1c6c3805111f1f1af85eR577)

File: src/raft/raft.rs (hunk 4 above)
Comment (mrmiywj) [inline src/raft/raft.rs]: https://github.com/coreos/etcd/pull/6039/files#diff-a8629a0eb6da1c6c3805111f1f1af85eR581

File: Cargo.lock (hunk 1 above)
Comment (mrmiywj) [inline Cargo.lock]: I'm not sure here. Since `Cargo.lock` is tracked in the git repo, I just added it.

**Comment by siddontang** [comment]
Hi @mrmiywj 

We find that etcd may have a bug for this implementation, I have told @xiang90 already. we should wait them fix it first then continue.

**Comment by mrmiywj** [comment]
Fine! We should wait them fix them!

**Comment by siddontang** [comment]
@mrmiywj 

etcd fixed https://github.com/coreos/etcd/pull/6129

**Comment by mrmiywj** [comment]
Hmm, it seems that the test fails.
`test raftstore::test_compact_log::test_node_compact_limit ... thread 'raftstore-1::test_node_compact_limit' panicked at 'attempt to subtract with overflow', src/raft/raft_log.rs:318
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread 'raftstore-3::test_node_compact_limit' panicked at 'attempt to subtract with overflow', src/raft/raft_log.rs:318
thread 'raftstore-5::test_node_compact_limit' panicked at 'attempt to subtract with overflow', src/raft/raft_log.rs:318
thread 'raftstore-4::test_node_compact_limit' panicked at 'attempt to subtract with overflow', src/raft/raft_log.rs:318
thread panicked while panicking. aborting.
error: Process didn't exit successfully: `/root/tikv/target/debug/tests-57d16572e02bef4a` (signal: 4, SIGILL: illegal instruction)`

**Comment by mrmiywj** [comment]
It seems the test fails.

> test raftstore::test_compact_log::test_node_compact_limit ... thread 'raftstore-1::test_node_compact_limit' panicked at 'attempt to subtract with overflow', src/raft/raft_log.rs:318
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread 'raftstore-3::test_node_compact_limit' panicked at 'attempt to subtract with overflow', src/raft/raft_log.rs:318
thread 'raftstore-5::test_node_compact_limit' panicked at 'attempt to subtract with overflow', src/raft/raft_log.rs:318
thread 'raftstore-4::test_node_compact_limit' panicked at 'attempt to subtract with overflow', src/raft/raft_log.rs:318
thread panicked while panicking. aborting.
error: Process didn't exit successfully: `/root/tikv/target/debug/tests-57d16572e02bef4a` (signal: 4, SIGILL: illegal instruction)

**Comment by siddontang** [comment]
Does this test failure occur in your local machine?

**Comment by mrmiywj** [comment]
Yep. It failed on the same test case.

**Comment by siddontang** [comment]
So I think this is introduced in your changes, maybe we should fix it first. 

seems that etcd doesn't handle overflow too. @xiang90

**Comment by mrmiywj** [comment]
But I'm not sure which caused it..

**Comment by siddontang** [comment]
you can use `cargo test test_node_compact_limit  -- --nocapture` and print the argument passing to slice or other places to see the overflow problem.

**Comment by siddontang** [comment]
PTAL @hhkbp2

**Comment by siddontang** [comment]
@mrmiywj 

Seem that etcd ignores this overflow error. see https://github.com/coreos/etcd/issues/6143

**Comment by hhkbp2** [comment]
Hi @mrmiywj,

There are 2 issues causing this PR fails to build:

1. function `soft_state()` should be deleted or marked as 'dead code' since the only usage of it in test case `test_raw_node_start()` is revised to `None`.

2. failure on test case `test_node_compact_limit` in `./tests/raftstore/test_compact_log.rs:149`. For those test nodes of empty peer, there still is the relationship of raft log:

    first index = last index + 1

and it will cause runtime panic in 

https://github.com/mrmiywj/tikv/blob/hide-campaign-rules/src/raft/raft_log.rs#L318

when calling `log.slice(6, 6)` in `raft.step()` because `first index` is 6 and `last index` is 5.
A bug report is issued to etcd raft, and i will keep an eye on the bug fix. After that bug fix is merged, we could merge this PR.

cc @siddontang

**Comment by mrmiywj** [comment]
@siddontang It passes tests on my linux machine now.

**Comment by siddontang** [comment]
LGTM 
PTAL @BusyJay @hhkbp2

**Comment by hhkbp2** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay

**Comment by BusyJay** [comment]
Please format your code by using `make all` or `make dev`.

**Comment by mrmiywj** [comment]
@BusyJay 
I just format the code I wrote. There are some formatting problems in the base code.
BTW, the format script in the Makefile does not work... May I should issue it?

**Comment by BusyJay** [comment]
Have you run `cargo install rustfmt`?

**Comment by mrmiywj** [comment]
@BusyJay 
Of course.

> root@avalon:~/tikv# rustfmt --write-mode=diff src/raft/raft.rs
Diff in src/raft/raft.rs at line 1269:
         self.prs = HashMap::with_capacity(meta.get_conf_state().get_nodes().len());⏎
         for &n in meta.get_conf_state().get_nodes() {⏎
             let next_idx = self.raft_log.last_index() + 1;⏎
-            let matched = if n == self.id {⏎
-                next_idx - 1⏎
-            } else {⏎
-                0⏎
-            };⏎
+            let matched = if n == self.id { next_idx - 1 } else { 0 };⏎
             self.set_progress(n, matched, next_idx);⏎
             info!("{} restored progress of {} [{:?}]",⏎
                   self.tag,⏎
I think the output could not be caught by `grep "Diff at line"`

File: src/raft/raft.rs [hunk 6]
```diff
@@ -664,6 +668,20 @@ impl<T: Storage> Raft<T> {
             if self.state == StateRole::Leader {
                 debug!("{} ignoring MsgHup because already leader", self.tag);
             } else {
+                let ents = self.raft_log
+                    .slice(self.raft_log.applied + 1,
+                           self.raft_log.committed + 1,
+                           u64::max_value())
```
Comment (BusyJay) [inline src/raft/raft.rs]: Please use `NO_LIMIT` instead.

File: src/raft/raft.rs [hunk 7]
```diff
@@ -664,6 +668,20 @@ impl<T: Storage> Raft<T> {
...
+                    .expect("unexpected error getting applied entries");
```
Comment (BusyJay) [inline src/raft/raft.rs]: s/applied/unapplied/

**Comment by BusyJay** [comment]
rest LGTM

**Comment by siddontang** [comment]
@BusyJay https://travis-ci.org/pingcap/tikv/jobs/151487794

**Pull request merged**